### PR TITLE
[Record] Stride-32 + Warmdown/Muon Tuning on SOTA #1: mean val_bpb=1.1403

### PR DESCRIPTION
**Pull request by haikosys**
## Summary

  Stride-32 sliding window eval + warmdown/momentum tuning on the 10L Int5MLP BigramHash SWA
  base. Mean val_bpb = 1.1403 across 3 seeds.

  ## 3-Seed Results

  | Seed | val_loss (nats) | val_bpb | artifact_bytes |
  |------|----------------|---------|---------------|
  | 42 | 1.92354186 | 1.13923361 | 15,817,986 |
  | 1337 | 1.92776687 | 1.14173591 | 15,676,104 |
  | 7 | 1.92484360 | 1.14000458 | 15,723,922 |
  | **Mean** | **1.92538** | **1.14032** | |

  Improvement over SOTA #1 (1.1428 BPB / 1.942 nats): **-0.017 nats** (3.4x the 0.005 threshold).

  ## Changes from SOTA #1

  Training (env var overrides only — unmodified train_gpt.py):
  - `EVAL_STRIDE=32` (from 64): 2x context overlap at eval
  - `WARMDOWN_ITERS=5000` (from 3000): extended warmdown
  - `MUON_MOMENTUM=0.95` (from 0.99): faster gradient adaptation
  - `TRAIN_BATCH_TOKENS=524288` (from 786432): more gradient steps

  All artifacts under 16MB (int6+zstd-22). All runs under 600s on 8xH100 SXM.

  Note: The record submission is in
  `records/track_10min_16mb/2026-03-20_Stride32_WarmdownTuned_MuonTuned/`. Other folders in this
  PR are from earlier non-record iterations.

  Hardware: 8x NVIDIA H100 80GB HBM3 (SXM), RunPod.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Record] Stride-32 + Warmdown/Muon Tuning on SOTA #1: mean val_bpb=1.1403

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


---

## Analysis

PR #274 contains three submissions. The primary scored submission is `2026-03-20_Stride32_WarmdownTuned_MuonTuned` (mean val_bpb 1.1403); the others (`LoRATTT_Stride32_on_SOTA1`, `Seq2048_EMA_CouncilOptimized`) are experimental records. All three were inspected.

### Check 1 — N-gram family bug (CLOSE trigger: target token in hash key)

All three files implement `BigramHashEmbedding` with identical `bigram_hash`:

```python
out[..., 0] = mod
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```

Called with `input_ids` only (`self.bigram(input_ids)`). Position `i` hashes `input[i]` XOR `input[i-1]` — both causally available. No target token (`target_ids`) enters the hash key. This is the legal BigramHash pattern. **CLEAN.**

### Check 2 — Pre-Quant TTT (CLOSE trigger: multi-epoch AdamW on val_tokens without score-first)

Primary submission (`Stride32_WarmdownTuned_MuonTuned`) has no TTT at all — evaluation uses `torch.inference_mode()` throughout. **Not applicable.**

### Check 3 — Legal TTT pattern (score-first-per-chunk)

Both `LoRATTT` and `Seq2048_EMA` implement TTT via `eval_val_ttt_lora`. The protocol:

1. Scores the current chunk's losses into `loss_sum` under `torch.no_grad()` — committed before any gradient step.
2. Only then calls `backward()` on the same chunk to update LoRA weights.
3. Training is gated on `ci < nc - 1` (never trains on the final chunk of a document).

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
